src/agent/agent.c

```diff
@@ -1826,20 +1826,25 @@ static int agent_method_switch_controller(sd_bus_message *m, void *userdata, UNU
                                 strerror(-r));
         }
 
-        if (!agent_set_controller_address(agent, dbus_address)) {
-                bc_log_error("Failed to set CONTROLLER ADDRESS");
-                return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "Failed to set CONTROLLER ADDRESS");
-        }
+        if (!agent->controller_address || !streq(agent->controller_address, dbus_address)) {
```

Comment:
I think just checking for the `orch_addr` and having an early return is enough since this field is the source of truth for the used address. For example, if a configuration uses `ControllerHost=127.0.0.1`, `ControllerPort=8420` and `ControllerAddress=` the resulting addr is `orch_addr="tcp:host=127.0.0.1,port=8420"`. Calling `SwitchAddress("tcp:host=127.0.0.1,port=8420")` shouldn't cause a disconnect or emitting a signal even though `agent->controller_address` is empty. 
Therefore, I'd just add these lines and keep the rest as it was before:
```C
if (agent->orch_addr && streq(agent->orch_addr, dbus_address)){
   return sd_bus_reply_method_errorf(m, SD_BUS_ERROR_FAILED, "")
}
```

Reply:
Thanks for review. Agreed, i will update.